HDFEventOutputer.cc

```diff
@@ -0,0 +1,204 @@
+#include "HDFEventOutputer.h"
+#include "OutputerFactory.h"
+#include "ConfigurationParameters.h"
+#include "UnrolledSerializerWrapper.h"
+#include "SerializerWrapper.h"
+#include "summarize_serializers.h"
+#include "lz4.h"
+#include <memory>
+#include <iostream>
+#include <cstring>
+#include <cmath>
+#include <set>
+
+using namespace cce::tf;
+using namespace cce::tf::pds;
+using product_t = std::vector<char>; 
+
+namespace {
+  template <typename T> 
+  void 
+  write_ds(hid_t gid, 
+           std::string const& dsname, 
+           std::vector<T> const& data) {
+    constexpr hsize_t ndims = 1;
+    auto dset = hdf5::Dataset::open(gid, dsname.c_str()); 
+    auto old_fspace = hdf5::Dataspace::get_space(dset);
+    hsize_t max_dims[ndims]; //= {H5S_UNLIMITED};
+    hsize_t old_dims[ndims]; //our datasets are 1D
+    H5Sget_simple_extent_dims(old_fspace, old_dims, max_dims);
+    //now old_dims[0] has existing length
+    //we need to extend by the size of data
+    hsize_t new_dims[ndims];
+    new_dims[0] = old_dims[0] + data.size();
+    hsize_t slab_size[ndims];
+    slab_size[0] = data.size();
+    dset.set_extent(new_dims);
+    auto new_fspace = hdf5::Dataspace::get_space (dset);
+    new_fspace.select_hyperslab(old_dims, slab_size);
+    auto mem_space = hdf5::Dataspace::create_simple(ndims, slab_size, max_dims);
+    dset.write<T>(mem_space, new_fspace, data); //H5Dwrite
+  }
+}
+
+HDFEventOutputer::HDFEventOutputer(std::string const& iFileName, unsigned int iNLanes, int iBatchSize, pds::Serialization iSerialization) : 
+  file_(hdf5::File::create(iFileName.c_str())),
+  maxBatchSize_{iBatchSize},
+  serializers_{std::size_t(iNLanes)},
+  serialization_{iSerialization},
+  serialTime_{std::chrono::microseconds::zero()},
+  parallelTime_{0}
+  {}
+
+HDFEventOutputer::~HDFEventOutputer() { }
+
+void HDFEventOutputer::setupForLane(unsigned int iLaneIndex, std::vector<DataProductRetriever> const& iDPs) {
+  auto& s = serializers_[iLaneIndex];
+  switch(serialization_) {
+  case pds::Serialization::kRoot:
+    {   s = SerializeStrategy::make<SerializeProxy<SerializerWrapper>>(); break; }
+  case pds::Serialization::kRootUnrolled:
+    {   s = SerializeStrategy::make<SerializeProxy<UnrolledSerializerWrapper>>(); break; }
+  }
+  s.reserve(iDPs.size());
+  offsetsAndBlob_.first.resize(iDPs.size()+1, 0);
+  for(auto const& dp: iDPs) {
+    s.emplace_back(dp.name(), dp.classType());
+  }
+  std::cout << "Setting up for Lanes\n"; 
```

Comment:
how about commenting this out?

Reply:
I just realized I use the name and type in writeFileHeader. I have moved it inside the if check right before writeHeader, so it gets called once.